### PR TITLE
feat(deps): Unblock the version of Angular to >= 2.3

### PR DESCRIPTION
**Pull request by hansl**
This includes using a private API not published.

Also added a SHA option when running e2e to install a SHA of Angular. This is the first 7 digits of an angular SHA that is on master.

### CODE REVIEW

File: packages/@ngtools/webpack/package.json [hunk 1]
```diff
@@ -31,9 +31,9 @@
     "source-map": "^0.5.6"
   },
   "peerDependencies": {
-    "@angular/compiler": "2.2.3",
-    "@angular/compiler-cli": "2.2.3",
-    "@angular/core": "2.2.3",
+    "@angular/compiler": "~2.3.1",
+    "@angular/compiler-cli": "~2.3.1",
+    "@angular/core": "~2.3.1",
     "@angular/tsc-wrapped": "0.4.0",
```
Comment (vicb) [inline packages/@ngtools/webpack/package.json]: 0.5.0 ?

File: packages/@ngtools/webpack/src/plugin.ts [hunk 2]
```diff
@@ -29,47 +26,22 @@ export interface AotPluginOptions {
...
-
-  toString() {
-    return `${this.path}#${this.className}`;
-  }
-
-  static fromString(entry: string): ModuleRoute {
-    const split = entry.split('#');
-    return new ModuleRoute(split[0], split[1]);
-  }
+  // We do not have an include because tsconfig can be used for that.
```
Comment (vicb) [inline packages/@ngtools/webpack/src/plugin.ts]: `// Use tsconfig to include` (positive form)

File: packages/@ngtools/webpack/package.json (hunk 1 above)
Comment (hansl) [inline packages/@ngtools/webpack/package.json]: Done.

File: packages/@ngtools/webpack/src/plugin.ts (hunk 2 above)
Comment (hansl) [inline packages/@ngtools/webpack/src/plugin.ts]: Done.

File: packages/@ngtools/webpack/src/plugin.ts [hunk 3]
```diff
@@ -93,7 +65,12 @@ export class AotPlugin implements Tapable {
   get compilerHost() { return this._compilerHost; }
   get compilerOptions() { return this._compilerOptions; }
   get done() { return this._donePromise; }
-  get entryModule() { return this._entryModule; }
+  get entryModule() {
+    const splitted = this._entryModule.split('#');
```
Comment (vicb) [inline packages/@ngtools/webpack/src/plugin.ts]: const {path, className} = this._entryModule.split('#');
return {path, className: className || 'default'};

File: packages/@ngtools/webpack/src/plugin.ts [hunk 4]
```diff
@@ -119,19 +96,59 @@ export class AotPlugin implements Tapable {
...
+    let fileNames = tsConfig.fileNames;
+    if (options.hasOwnProperty('exclude')) {
+      let exclude: string[] = typeof options.exclude == 'string'
+          ? [options.exclude as string] : (options.exclude as string[]);
+
+      exclude.forEach((pattern: string) => {
+        const basePathPattern = '(' + basePath.replace(/\\/g, '/')
+            .replace(/[\-\[\]\/{}()+?.\\^$|*]/g, '\\$&') + ')?';
+        pattern = pattern
+          // Replace windows path separators to forward slashes.
```
Comment (vicb) [inline packages/@ngtools/webpack/src/plugin.ts]: with

Comment (hansl) [inline packages/@ngtools/webpack/src/plugin.ts]: Donezzzo

File: packages/angular-cli/blueprints/ng2/files/package.json [hunk 5]
```diff
@@ -13,21 +13,21 @@
...
+    "@angular/common": "~2.3.1",
+    "@angular/compiler": "~2.3.1",
+    "@angular/core": "~2.3.1",
+    "@angular/forms": "~2.3.1",
+    "@angular/http": "~2.3.1",
+    "@angular/platform-browser": "~2.3.1",
+    "@angular/platform-browser-dynamic": "~2.3.1",
     "@angular/router": "3.2.3",
     "core-js": "^2.4.1",
     "rxjs": "5.0.0-beta.12",
```
Comment (vicb) [inline packages/angular-cli/blueprints/ng2/files/package.json]: update?

File: packages/angular-cli/blueprints/ng2/files/package.json [hunk 6]
```diff
@@ -13,21 +13,21 @@
...
     "ts-helpers": "^1.1.1",
     "zone.js": "^0.6.23"
```
Comment (vicb) [inline packages/angular-cli/blueprints/ng2/files/package.json]: update

File: packages/angular-cli/commands/build.ts [hunk 7]
```diff
@@ -59,6 +60,10 @@ const BuildCommand = Command.extend({
     }
 
     const project = this.project;
+
+    // Check angular version.
+    Version.assertAngularVersionIs2_3_1OrBetter(project.root);
```
Comment (vicb) [inline packages/angular-cli/commands/build.ts]: I would say "Higher" :)

File: packages/angular-cli/upgrade/version.ts [hunk 8]
```diff
@@ -45,6 +45,8 @@ export class Version {
   isReleaseCandidate() { return this.qualifier == 'rc'; }
   isKnown() { return this._version !== null; }
 
+  isLocal() { return this.isKnown() ? path.isAbsolute(this._version) : false; }
```
Comment (vicb) [inline packages/angular-cli/upgrade/version.ts]: return this.isKnown() && path.isAbsolute(this._version);

File: packages/angular-cli/blueprints/ng2/files/package.json (hunk 5 above)
Comment (hansl) [inline packages/angular-cli/blueprints/ng2/files/package.json]: done.

File: packages/angular-cli/blueprints/ng2/files/package.json (hunk 6 above)
Comment (hansl) [inline packages/angular-cli/blueprints/ng2/files/package.json]: done.

File: packages/angular-cli/commands/build.ts (hunk 7 above)
Comment (hansl) [inline packages/angular-cli/commands/build.ts]: Good point.

File: packages/angular-cli/upgrade/version.ts [hunk 9]
```diff
@@ -91,6 +93,36 @@ export class Version {
...
+      if (v.isLocal()) {
+        console.warn(yellow('Using a local version of angular. Proceeding with care...'));
+      } else {
+        if (v.major != 2 || ((v.minor == 3 && v.patch == 0) || v.minor < 3)) {
+          console.error(bold(red(stripIndents`
+            This version of CLI is only compatible with angular version 2.3.1 or better. Please
+            upgrade your angular version, e.g. by running:
+            
+            npm install @angular/core@latest
+          ` + '\n')));
```
Comment (vicb) [inline packages/angular-cli/upgrade/version.ts]: nit `...` + '\n' ?

File: packages/angular-cli/upgrade/version.ts (hunk 8 above)
Comment (hansl) [inline packages/angular-cli/upgrade/version.ts]: Done.

**Comment by vicb** [comment]
lgtm with a few comments

File: packages/angular-cli/upgrade/version.ts (hunk 9 above)
Comment (hansl) [inline packages/angular-cli/upgrade/version.ts]: `stripIndents` remove `\n`s so we have to concatenate the string.

**Comment by hansl** [comment]
Thanks a lot @vicb !

**Pull request merged**